### PR TITLE
[FIX] mass_mailing: apply design changes test mailing

### PR DESCRIPTION
**Pull request by Mtaylorr**
Issue:
======
Email doesn't have applied design changes (format , color ..)

Steps to reproduce the issue:
=============================
- Go to Email marketing
- Create a new one
- Add a subject, mailing list, and choose any template that have some
  blocks
- Go to design tab in editor and change any color or size of something
- Click save and test sending the email
- The email doesn't have the changes applied

Origin of the issue:
====================
When first rendering , CssRules are calculated using the first version
of the template and styles. When we update the design so the css rules,
they are not applied in the inlineHtml since it uses the old cssRules.

Solution:
==========
We only use `_rulesCache` defined in `wysiwyg` and we already handle
everything in `toInline` from calculating the `cssRules` to saving them
in the cache, so we just pass `undefined` as cssRules and let it take
care of everything.

task-3289131

### CODE REVIEW

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 1]
```diff
@@ -129,8 +129,8 @@ export class MassMailingHtmlField extends HtmlField {
             // Wait for the css and images to be loaded.
             await iframePromise;
             const editableClone = iframe.contentDocument.querySelector('.note-editable');
-            this.cssRules = this.cssRules || getCSSRules($editable[0].ownerDocument);
-            await toInline($(editableClone), this.cssRules, $(iframe));
+            const cssRules = this.cssRules || getCSSRules($editable[0].ownerDocument);
+            await toInline($(editableClone), cssRules, $(iframe));
```
Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:133]: This completely disables the cache, thus greatly slowing down the operations of mass_mailing that trigger a convert inline. I quickly told @Zynton about the issue and he suggested to invalidate the cache when something is changed in the design tab. That would be a better way to fix the issue so that we can keep the caching feature.

Comment (Mtaylorr) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: @dmo-odoo  , I changed the fix , can you take a look ? Thanks

Comment (dmo-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: Unless I'm mistaken, I don't think this is what @Zynton had in mind. Maybe he can chime in?

Comment (Zynton) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: Indeed, I meant to just invalidate the cache whenever a change is made in the design tab.

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 2]
```diff
@@ -129,8 +129,7 @@ export class MassMailingHtmlField extends HtmlField {
             // Wait for the css and images to be loaded.
             await iframePromise;
             const editableClone = iframe.contentDocument.querySelector('.note-editable');
-            this.cssRules = this.cssRules || getCSSRules($editable[0].ownerDocument);
-            await toInline($(editableClone), this.cssRules, $(iframe));
+            await toInline($(editableClone), this.wysiwyg._rulesCache, $(iframe));
```
Comment (Zynton) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:132]: I believe you're correct that `this.cssRules` shouldn't have been defined here. I notice that in 15.0, this instance variable is _passed_ to `toInline` in two places, _but never defined_. So it's effectively useless. Then we [defined it](https://github.com/odoo/odoo/commit/1173733ba7c33abf079d668ea219e316401352d3#diff-0b4e81825324dc36e89543232f3f8f8704acddc4ec21031c6ccea78ee93b00eeR127), and we end up with two distinct caches: `HtmlField.cssRules` and `Wysiwyg._rulesCache`.
Now, `toInline` already does this:
```js
cssRules = cssRules || wysiwyg && wysiwyg._rulesCache;
if (!cssRules) {
    cssRules = getCSSRules(doc);
    if (wysiwyg) {
        wysiwyg._rulesCache = cssRules;
    }
}
```
so _as long as `toInline` can find `wysiwyg`_, passing it `this.wysiwyg._rulesCache` should not have any effect.

In conclusion, I would remove any reference to `this.cssRules` anywhere in the codebase and replace it with `undefined`. Then check if everything works. If it doesn't it's likely because `toInline` can't find `wysiwyg` at some point. But I suspect it will work :-)

```suggestion
            await toInline($(editableClone), undefined, $(iframe));
```

cc @dmo-odoo

Comment (Mtaylorr) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: @Zynton Thanks for the suggestion, I updated the change , can you take a look when you have time ? thanks

Comment (Zynton) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: LGTM. @dmo-odoo ?

Comment (dmo-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: Did one of you test that the cache was properly used when it should? That is we should not recompute the rules twice when calling `toInline` twice if we didn't change anything in the design tab.

Comment (Mtaylorr) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: I didn't think about this , I will test it and get back to you

Comment (Mtaylorr) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: @dmo-odoo , I changed the fix here for stable, and I will update it in master to pass options into `toInline`

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 3]
```diff
@@ -129,8 +129,13 @@ export class MassMailingHtmlField extends HtmlField {
             // Wait for the css and images to be loaded.
             await iframePromise;
             const editableClone = iframe.contentDocument.querySelector('.note-editable');
-            this.cssRules = this.cssRules || getCSSRules($editable[0].ownerDocument);
-            await toInline($(editableClone), this.cssRules, $(iframe));
+            // editableClone doesn't have wysiwyg in its data
+            // we use this hacky fix to assign it now so we can get it later in toInline
+            // from the editable
+            // TODO in master : Update the function toInline to accept an options dictionary
+            // as paraemeter.
```
Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:136]: Reworded to wrap at 80 chars:
```suggestion
            // The jQuery data are lost because of the cloning operations above.
            // The hacky fix for stable is to simply add it back manually.
            // TODO in master: Update toInline to use an options parameter.
```
Other than that, great catch ! 👍

Comment (dmo-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_html_field.js:136]: Reworded to wrap at 80 chars:
```suggestion
            // The jQuery data are lost because of the cloning operations above.
            // The hacky fix for stable is to simply add it back manually.
            // TODO in master: Update toInline to use an options parameter.
```
Anyway, great catch ! 👍

**Comment by Mtaylorr** [comment]
@dmo-odoo  , should we merge this ?

**Comment by dmo-odoo** [comment]
@robodoo r+
